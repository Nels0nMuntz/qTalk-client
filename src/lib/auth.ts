import { NextAuthOptions, getServerSession } from 'next-auth';
import GoogleProvider from 'next-auth/providers/google';
import CredentialsProvider from 'next-auth/providers/credentials';
import { PrismaAdapter } from '@auth/prisma-adapter';
import { nanoid } from 'nanoid';
import { compare } from 'bcrypt';
import { db } from './db';

export const authoOptions: NextAuthOptions = {
  adapter: PrismaAdapter(db),
  session: {
    strategy: 'jwt',
  },
  pages: {
    signIn: '/sign-in',
  },
  secret: process.env.NEXTAUTH_SECRET,
  // debug: process.env.NODE_ENV === 'development',
  providers: [
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID!,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET!,
    }),
    CredentialsProvider({
      type: 'credentials',
      name: 'credentials',
      credentials: {
        email: { label: 'Email', type: 'text' },
        password: { label: 'Password', type: 'text' },
      },
      async authorize(credentials) {
        if (!credentials?.email || !credentials.password) {
          return null;
        }
        const userbyEmail = await db.user.findFirst({
          where: {
            email: credentials.email,
          },
        });

        if (!userbyEmail || !userbyEmail.password) {
          return null;
        }

        const isPasswordValid = await compare(
          credentials.password,
          userbyEmail.password,
        );

        if (!isPasswordValid) {
          return null;
        }

        return userbyEmail;
      },
    }),
  ],
  callbacks: {
    async session({ token, session }) {
      if (token) {
        session.user.id = token.id;
        session.user.name = token.name;
        session.user.email = token.email;
        session.user.image = token.picture;
        session.user.username = token.username;
      }

      return session;
    },

    async jwt({ token, user }) {
      const dbUser = await db.user.findFirst({
        where: {
          email: token.email,
        },
      });

      if (!dbUser) {
        token.id = user!.id;
        return token;
      }

      if (!dbUser.username) {
        await db.user.update({
          where: {
            id: dbUser.id,
          },
          data: {
            username: nanoid(10),
          },
        });
      }

      return {
        id: dbUser.id,
        name: dbUser.name,
        email: dbUser.email,
        picture: dbUser.image,
        username: dbUser.username,
      };
    },
  },
};

export const getAuthSession = () => getServerSession(authoOptions);
