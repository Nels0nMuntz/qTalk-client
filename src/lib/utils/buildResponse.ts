export const buildResponse = (body: string | object, init?: ResponseInit) => {
  const responseBody =
    typeof body === 'string'
      ? JSON.stringify({ message: body })
      : JSON.stringify(body);
  return new Response(responseBody, init);
};
