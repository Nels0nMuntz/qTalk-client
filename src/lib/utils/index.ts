export * from './buildResponse';
export * from './utils';
export * from './notify';
