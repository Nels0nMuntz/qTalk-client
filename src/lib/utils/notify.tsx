import toast, { ToastOptions, Toast } from 'react-hot-toast';
import { CheckCircle, XCircle, X } from 'lucide-react';
import React from 'react';

type NotifyVariants = 'success' | 'error';

const icons: Record<NotifyVariants, JSX.Element> = {
  success: <CheckCircle color="#a3e635" />,
  error: <XCircle color="#f87171" />,
};

export const notify = ({
  title,
  description,
  render,
  variant,
  options,
}: {
  title?: string;
  description?: string;
  render?: (ctx: Toast) => React.ReactNode;
  variant: 'success' | 'error';
  options?: ToastOptions;
}) => {
  const Icon = icons[variant];
  return toast(
    (t) => (
      <div className="flex items-start">
        <div className="shrink-0 mr-4">{Icon}</div>
        <div className="flex flex-col gap-y-1">
          <span className="text-sm font-semibold">{title}</span>
          <span className="text-sm" onClick={() => toast.dismiss(t.id)}>
            {description}
          </span>
          {render && render(t)}
        </div>
        <div className="ml-4 shrink-0">
          <button
            className="text-zinc-500 hover:text-zinc-700 transition-all rounded-md focus:outline focus:outline-offset-1 focus:outline-2 focus:outline-indigo-500"
            type="button"
            onClick={() => toast.dismiss(t.id)}
          >
            <span className="sr-only">Close</span>
            <X size={20} />
          </button>
        </div>
      </div>
    ),
    options,
  );
};
