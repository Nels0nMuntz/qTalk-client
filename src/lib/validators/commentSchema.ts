import { z } from 'zod';

export const commentSchema = z.object({
  postId: z.string(),
  replyToId: z.string().optional(),
  text: z.string(),
});

export type CreateCommentPayload = z.infer<typeof commentSchema>;
