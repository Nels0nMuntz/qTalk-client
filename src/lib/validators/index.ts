export * from './signInFormSchema';
export * from './signUpFormSchema';
export * from './subtalkSchema';
