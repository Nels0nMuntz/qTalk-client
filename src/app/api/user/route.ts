import { db } from '@/lib/db';
import { buildResponse } from '@/lib/utils';
import { signUpFormSchema } from '@/lib/validators';
import { hash } from 'bcrypt';
import { z } from 'zod';

export async function POST(req: Request) {
  try {
    const body = await req.json();
    const { email, username, password } = signUpFormSchema.parse(body);

    const userByEmail = await db.user.findFirst({
      where: {
        email,
      },
    });
    const userByUsername = await db.user.findFirst({
      where: {
        username,
      },
    });
    if (userByEmail || userByUsername) {
      return buildResponse('Email or username are taken', { status: 422 });
    }

    const hashedPassword = await hash(password, 10);
    await db.user.create({
      data: {
        email,
        username,
        password: hashedPassword,
      },
    });
    return buildResponse('User was created', { status: 201 });
  } catch (error) {
    console.log(error);

    if (error instanceof z.ZodError) {
      return new Response(error.message, { status: 422 });
    }

    return buildResponse('Internal server error', { status: 500 });
  }
}
