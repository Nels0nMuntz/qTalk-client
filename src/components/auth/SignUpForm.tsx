'use client';

import React from 'react';
import { useForm } from 'react-hook-form';
import { zodResolver } from '@hookform/resolvers/zod';
import {
  Form,
  FormControl,
  FormField,
  FormItem,
  FormLabel,
  FormMessage,
} from '../ui/form';
import { Input } from '../ui/input';
import { SignUpFormSchema, signUpFormSchema } from '@/lib/validators';
import { Button } from '../ui/button';
import { useRouter } from 'next/navigation';
import { notify } from "@/lib/utils";

export default function SignUpForm() {
  const router = useRouter();
  const [isSubmitting, setIsSubmitting] = React.useState(false);
  const form = useForm<SignUpFormSchema>({
    resolver: zodResolver(signUpFormSchema),
    defaultValues: {
      username: '',
      email: '',
      password: '',
      confirmPassword: '',
    },
  });

  const onSubmit = async (values: SignUpFormSchema) => {
    try {
      setIsSubmitting(true);
      const response = await fetch('/api/user', {
        method: 'POST',
        body: JSON.stringify(values),
        headers: {
          "Content-Type": "application/json",
        },
      });
      const data = await response.json();
      if (response.status === 201) {
        notify({ title: data.message, variant: 'error' });
        router.push('/sign-in');
      } else {
        notify({ title: data.message, variant: 'error' });
      }
    } catch (error) {
      console.log(error);
    } finally {
      setIsSubmitting(false);
    }
  };

  return (
    <Form {...form}>
      <form
        className="space-y-4 text-left"
        onSubmit={form.handleSubmit(onSubmit)}
      >
        <FormField
          control={form.control}
          name="username"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Username</FormLabel>
              <FormControl>
                <Input placeholder="your name" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="email"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Email</FormLabel>
              <FormControl>
                <Input placeholder="your@email.com" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="password"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Password</FormLabel>
              <FormControl>
                <Input placeholder="at least 6 characters" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <FormField
          control={form.control}
          name="confirmPassword"
          render={({ field }) => (
            <FormItem className="space-y-1">
              <FormLabel>Confirm Password</FormLabel>
              <FormControl>
                <Input placeholder="the same as above" {...field} />
              </FormControl>
              <FormMessage />
            </FormItem>
          )}
        />
        <Button
          className="w-full text-white bg-rose-400 rounded-lg hover:bg-rose-500 transition-colors"
          isLoading={isSubmitting}
        >
          Get Started
        </Button>
      </form>
    </Form>
  );
}
