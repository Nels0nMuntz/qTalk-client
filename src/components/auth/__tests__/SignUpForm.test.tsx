import { render, screen } from '@testing-library/react';
import user from '@testing-library/user-event';
import SignUpForm from '../SignUpForm';
import { SignUpFormSchema } from '@/lib/validators';
import { notify } from '@/lib/utils/notify';

const pushMock = jest.fn()
jest.mock('next/navigation', () => ({
  useRouter: () => ({
    push: pushMock,
  }),
}));

jest.mock('@/lib/utils/notify.tsx', () => ({
  notify: jest.fn(),
}));

const queryUsernameHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:nth-child(1) > input + p');
};
const queryEmailHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:nth-child(2) > input + p');
};
const queryPasswordHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:nth-child(3) > input + p');
};
const queryConfirmPasswordHelperText = (container: HTMLElement) => {
  return container.querySelector('form > div:nth-child(4) > input + p');
};

const submitForm = async (formData: SignUpFormSchema) => {
  render(<SignUpForm />);

  const usernameInput = await screen.findByLabelText('Username');
  const emailInput = await screen.findByLabelText('Email');
  const passwordInput = await screen.findByLabelText('Password');
  const confirmPasswordInput = await screen.findByLabelText('Confirm Password');
  const submitButton = await screen.findByRole('button', {
    name: /get started/i,
  });

  await user.type(usernameInput, formData.username);
  await user.type(emailInput, formData.email);
  await user.type(passwordInput, formData.password);
  await user.type(confirmPasswordInput, formData.confirmPassword);
  await user.click(submitButton);
};

describe('SignUpForm component', () => {
  it('renders correctly', async () => {
    const { container } = render(<SignUpForm />);

    const form = await screen.findByRole('form', { name: 'signup form' });

    expect(form).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('renders Username validation errors correctly', async () => {
    const { container } = render(<SignUpForm />);

    const usernameInput = await screen.findByLabelText('Username');
    const submitButton = await screen.findByRole('button', {
      name: /get started/i,
    });

    expect(usernameInput).toHaveValue('');
    await user.click(submitButton);
    expect(queryUsernameHelperText(container)).toHaveTextContent(
      'String must contain at least 2 character(s)',
    );

    await user.type(usernameInput, 'U');
    expect(queryUsernameHelperText(container)).toHaveTextContent(
      'String must contain at least 2 character(s)',
    );

    await user.clear(usernameInput);
    await user.type(
      usernameInput,
      'more then 50 character more then 50 character more then 50 character',
    );
    expect(queryUsernameHelperText(container)).toHaveTextContent(
      'String must contain at most 50 character(s)',
    );

    await user.clear(usernameInput);
    await user.type(usernameInput, 'Mark');
    expect(queryUsernameHelperText(container)).toBeFalsy();
  });

  it('renders emain validation errors correctly', async () => {
    const { container } = render(<SignUpForm />);

    const emailInput = await screen.findByLabelText('Email');
    const submitButton = await screen.findByRole('button', {
      name: /get started/i,
    });

    expect(emailInput).toHaveValue('');
    await user.click(submitButton);
    expect(queryEmailHelperText(container)).toHaveTextContent('Invalid email');

    await user.type(emailInput, 'wrong@email');
    expect(queryEmailHelperText(container)).toHaveTextContent('Invalid email');

    await user.clear(emailInput);
    await user.type(emailInput, 'right@gmail.com');
    expect(queryEmailHelperText(container)).toBeFalsy();
  });

  it('renders password validation errors correctly', async () => {
    const { container } = render(<SignUpForm />);

    const passwordInput = await screen.findByLabelText('Password');
    const submitButton = await screen.findByRole('button', {
      name: /get started/i,
    });

    expect(passwordInput).toHaveValue('');
    await user.click(submitButton);
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at least 6 character(s)',
    );

    await user.type(passwordInput, 'd7a92');
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at least 6 character(s)',
    );

    await user.clear(passwordInput);
    await user.type(passwordInput, 'd7a92f5c4e5d9a3b4f7c5');
    expect(queryPasswordHelperText(container)).toHaveTextContent(
      'String must contain at most 20 character(s)',
    );

    await user.clear(passwordInput);
    await user.type(passwordInput, 'd7a92f5c4e5d9a3b4f7c');
    expect(queryPasswordHelperText(container)).toBeFalsy();
  });

  it('renders confirm password field validation errors correctly', async () => {
    const { container } = render(<SignUpForm />);

    const passwordInput = await screen.findByLabelText('Password');
    const confirmPasswordInput =
      await screen.findByLabelText('Confirm Password');
    const submitButton = await screen.findByRole('button', {
      name: /get started/i,
    });

    expect(passwordInput).toHaveValue('');
    expect(confirmPasswordInput).toHaveValue('');

    await user.type(passwordInput, 'd7a92f5c4e5d9a3b4f7c');
    await user.click(submitButton);
    expect(queryConfirmPasswordHelperText(container)).toHaveTextContent(
      'Passwords must match!',
    );

    await user.type(confirmPasswordInput, 'd7a92f5c4e5d9a3b4f7c'.toUpperCase());
    expect(queryConfirmPasswordHelperText(container)).toHaveTextContent(
      'Passwords must match!',
    );

    await user.clear(confirmPasswordInput);
    await user.type(confirmPasswordInput, 'd7a92f5c4e5d9a3b4f7c');
    expect(queryConfirmPasswordHelperText(container)).toBeFalsy();
  });

  it('submits form with form values', async () => {
    global.fetch = jest.fn(() => {
      return Promise.resolve({
        json: () => Promise.resolve({}),
        status: 201,
      });
    }) as jest.Mock;

    const values = {
      username: 'Jenna',
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      confirmPassword: 'jennajenna',
    };

    await submitForm(values);

    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith('/api/user', {
      method: 'POST',
      body: JSON.stringify(values),
      headers: {
        'Content-Type': 'application/json',
      },
    });
  });

  it('redirects to sign-in page after user was successfully created', async () => {
    const responseMessage = 'User created successfully';
    global.fetch = jest.fn(() => {
      return Promise.resolve({
        json: () => Promise.resolve({ message: responseMessage }),
        status: 201,
      });
    }) as jest.Mock;

    const values = {
      username: 'Jenna',
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      confirmPassword: 'jennajenna',
    };

    await submitForm(values);

    expect(notify).toHaveBeenCalledWith({
      title: responseMessage,
      variant: 'success',
    })
    expect(pushMock).toHaveBeenCalledWith('/sign-in');
  });

  it('shows error notification after signup request faild', async () => {
    const responseMessage = 'Error';
    global.fetch = jest.fn(() => {
      return Promise.resolve({
        json: () => Promise.resolve({ message: responseMessage })
      })
    }) as jest.Mock;

    const values = {
      username: 'Jenna',
      email: 'jenna@jenna.com',
      password: 'jennajenna',
      confirmPassword: 'jennajenna',
    };

    await submitForm(values);

    expect(notify).toHaveBeenCalledWith({
      title: responseMessage,
      variant: 'error',
    })
  });
});
