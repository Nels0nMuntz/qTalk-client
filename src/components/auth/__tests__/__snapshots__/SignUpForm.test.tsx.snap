// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm component renders correctly 1`] = `
<div>
  <form
    aria-label="signup form"
    class="space-y-4 text-left"
  >
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r0:-form-item"
      >
        Username
      </label>
      <input
        aria-describedby=":r0:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r0:-form-item"
        name="username"
        placeholder="your name"
        value=""
      />
    </div>
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r1:-form-item"
      >
        Email
      </label>
      <input
        aria-describedby=":r1:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r1:-form-item"
        name="email"
        placeholder="your@email.com"
        value=""
      />
    </div>
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r2:-form-item"
      >
        Password
      </label>
      <input
        aria-describedby=":r2:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r2:-form-item"
        name="password"
        placeholder="at least 6 characters"
        value=""
      />
    </div>
    <div
      class="space-y-1"
    >
      <label
        class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
        for=":r3:-form-item"
      >
        Confirm Password
      </label>
      <input
        aria-describedby=":r3:-form-item-description"
        aria-invalid="false"
        class="flex h-10 w-full rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background file:border-0 file:bg-transparent file:text-sm file:font-medium placeholder:text-muted-foreground focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50"
        id=":r3:-form-item"
        name="confirmPassword"
        placeholder="the same as above"
        value=""
      />
    </div>
    <button
      class="inline-flex items-center justify-center text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 h-10 px-4 py-2 w-full text-white bg-rose-400 rounded-lg hover:bg-rose-500 transition-colors"
    >
      Get Started
    </button>
  </form>
</div>
`;
